File: python/protein-translation/protein_translation.py
def proteins(strand):
  m = {
    'AUG': 'Methionine',
    'UUU': 'Phenylalanine', 'UUC': 'Phenylalanine',
    'UUA': 'Leucine', 'UUG': 'Leucine',
    'UCU': 'Serine', 'UCC': 'Serine', 'UCA': 'Serine', 'UCG': 'Serine',
    'UAU': 'Tyrosine', 'UAC': 'Tyrosine',
    'UGU': 'Cysteine', 'UGC': 'Cysteine',
    'UGG': 'Tryptophan',
    'UAA': 'STOP', 'UAG': 'STOP', 'UGA': 'STOP',
  }
  proteins = []
  index = 0
  while index < len(strand) - 2:
    if m[strand[index:index+3]] == 'STOP':
      return proteins
    if m[strand[index:index+3]]:
      proteins.append(m[strand[index:index+3]])
    index += 3

  return proteins

File: csharp/protein-translation/ProteinTranslation.cs
using System;
using System.Collections.Generic;

public static class ProteinTranslation
{
    private static Dictionary<string, string[]> givenProteinMap = new Dictionary<string, string[]>(){
        { "Methionine", new string[]{"AUG"} },
        { "Phenylalanine", new string[]{ "UUU", "UUC" } },
        { "Leucine", new string[]{ "UUA", "UUG" } },
        { "Serine", new string[]{ "UCU", "UCC", "UCA", "UCG" } },
        { "Tyrosine", new string[]{ "UAU", "UAC" } },
        { "Cysteine", new string[]{ "UGU", "UGC" } },
        { "Tryptophan", new string[]{ "UGG"} },
        { "STOP", new string[]{ "UAA", "UAG", "UGA" } },
    };
    private static Dictionary<string, string> derivedProteinMap = new Dictionary<string, string>();

    static ProteinTranslation() {
        // convert given protein map to one-to-one map
        foreach (KeyValuePair<string, string[]> entry in givenProteinMap) {
            foreach (string codon in entry.Value) {
                derivedProteinMap.Add(codon, entry.Key);
            }
        }
    }
    public static string[] Proteins(string strand)
    {
        var res = new List<string>();
        for (int i = 0; i < strand.Length - 2; i += 3) {
            var protein = derivedProteinMap[strand.Substring(i, 3)];
            if (protein == "STOP") {
                break;
            }
            res.Add(protein);
        }
        return res.ToArray();
    }
}